src/conf/domain_conf.c

```diff
@@ -19897,19 +19901,69 @@ virDomainDefParseXML(xmlXPathContextPtr ctxt,
                                                        ctxt,
                                                        nodes[i],
                                                        flags);
+
         if (!chr)
             return NULL;
 
-        if (chr->target.port == -1) {
+        def->serials[def->nserials++] = chr;
+
+        // Giving precedence to the isa-serial device since
+        // only limited ports can be used for such devices.
+        if (chr->targetType == VIR_DOMAIN_CHR_SERIAL_TARGET_MODEL_ISA_SERIAL) {
+            // Taking the isa serial deivces to start of the array.
+            for (j = def->nserials; j > isa_serial_count; j--)
```

Comment:
def->nserials is only for ISA serial? What if the domain definition have PCI or USB serial devices? Is that counted separately?

Reply:
it gives the count of all the serial devices.

Reply:
so basically for this particular iteration def->nserials indicates total number of serial devices covered till now plus the one on which the iteration is right now processing. So it will be equalt to 'i'
def->nserials = i for the ith iteration

Reply:
I'm wondering if there is a better way to do this instead of shifting everything inside a loop - but don't have any real suggestions for this

Reply:
happy to get suggestions.

Reply:
let me know if this makes sense, not necessary but your choice.
have two arrays defined
isa_serial[]
non_isa_serial[]
do first loop over all devices to separate out isa and non-isa in different array
run loop over isa_serial array with your logic to populate def->serials
use already existing loop or logic on non_isa to populate def->serials

Reply:
thats what I am doing here without the use of extra arrays. I am shifting all the isa-serial devices to the top applying my logic on them and then for rest of the devices applying the original logic.

Reply:
above approach will leave a lot of overhead of freeing up the memory as we dont know what should be the size of the two arrays.

Reply:
both array can be of size n. These are just pointers. Even if you do not want different arrays keep same local array. Atleast one loop can be used only for isa_serial. Next loop keep complete code as it is but just do it for non-isa.

Reply:
but then that will leave the change of overlapping ports. Which anyway we dont want whatever device it is.

Reply:
why overlapping ports.

Reply:
for isa you will be checking it is not used before and for non-isa maxport logic will give port which is not assigned yet?

Reply:
like if the given order is :

1. isa-serial 
2. pci-serial 
3. pci-serial 
4. pci-serial 
5. pci-serial 3

then we will get over lapping for pci device 3 and 5

Reply:
Please keep your existing change if other are also ok with this. I am ok with both.

Reply:
okay

Reply:
that overlapping is there even now.
1. pci-serial 250
2. pci-serial 250
will these give error as of now?

Reply:
sorry i am not sure what is max support port number, i just wanted to take some high value.

Reply:
yes

Reply:
right now for non isa-serial devices we dont have any limit.